File: src/BigInt.php
<?php
declare(strict_types=1);

namespace Pastafari\Stage01;

use InvalidArgumentException;
use RuntimeException;

final class BigInt
{
    private const BASE = 1000000000;
    private const BASE_DIGITS = 9;

    private int $sign;
    /** @var list<int> */
    private array $limbs;

    /** @param list<int> $limbs */
    private function __construct(int $sign, array $limbs)
    {
        while (count($limbs) > 1 && $limbs[array_key_last($limbs)] === 0) {
            array_pop($limbs);
        }
        if ($limbs === [] || (count($limbs) === 1 && $limbs[0] === 0)) {
            $this->sign = 0;
            $this->limbs = [0];
            return;
        }
        $this->sign = $sign < 0 ? -1 : 1;
        $this->limbs = array_values($limbs);
    }

    public static function zero(): self
    {
        return new self(0, [0]);
    }

    public static function one(): self
    {
        return new self(1, [1]);
    }

    public static function fromInt(int $value): self
    {
        return self::fromString((string)$value);
    }

    public static function fromString(string $value): self
    {
        $value = trim($value);
        if (!preg_match('/^[+-]?\d+$/', $value)) {
            throw new InvalidArgumentException('BIGINT_FORMAT');
        }

        $sign = 1;
        if ($value[0] === '-') {
            $sign = -1;
            $value = substr($value, 1);
        } elseif ($value[0] === '+') {
            $value = substr($value, 1);
        }

        $value = ltrim($value, '0');
        if ($value === '') {
            return self::zero();
        }

        $limbs = [];
        for ($end = strlen($value); $end > 0; $end -= self::BASE_DIGITS) {
            $start = max(0, $end - self::BASE_DIGITS);
            $limbs[] = (int)substr($value, $start, $end - $start);
        }
        return new self($sign, $limbs);
    }

    public function isZero(): bool
    {
        return $this->sign === 0;
    }

    public function sign(): int
    {
        return $this->sign;
    }

    public function abs(): self
    {
        return new self($this->isZero() ? 0 : 1, $this->limbs);
    }

    public function neg(): self
    {
        return new self(-$this->sign, $this->limbs);
    }

    public function compare(self $other): int
    {
        if ($this->sign !== $other->sign) {
            return $this->sign <=> $other->sign;
        }
        if ($this->sign === 0) {
            return 0;
        }
        $cmp = self::compareAbsLimbs($this->limbs, $other->limbs);
        return $this->sign > 0 ? $cmp : -$cmp;
    }

    public function equals(self $other): bool
    {
        return $this->compare($other) === 0;
    }

    public function add(self $other): self
    {
        if ($this->sign === 0) {
            return new self($other->sign, $other->limbs);
        }
        if ($other->sign === 0) {
            return new self($this->sign, $this->limbs);
        }
        if ($this->sign === $other->sign) {
            return new self($this->sign, self::addAbsLimbs($this->limbs, $other->limbs));
        }

        $cmp = self::compareAbsLimbs($this->limbs, $other->limbs);
        if ($cmp === 0) {
            return self::zero();
        }
        if ($cmp > 0) {
            return new self($this->sign, self::subAbsLimbs($this->limbs, $other->limbs));
        }
        return new self($other->sign, self::subAbsLimbs($other->limbs, $this->limbs));
    }

    public function sub(self $other): self
    {
        return $this->add($other->neg());
    }

    public function mul(self $other): self
    {
        if ($this->isZero() || $other->isZero()) {
            return self::zero();
        }
        $out = array_fill(0, count($this->limbs) + count($other->limbs) + 1, 0);
        foreach ($this->limbs as $i => $a) {
            $carry = 0;
            foreach ($other->limbs as $j => $b) {
                $idx = $i + $j;
                $cur = $out[$idx] + $a * $b + $carry;
                $out[$idx] = $cur % self::BASE;
                $carry = intdiv($cur, self::BASE);
            }
            $idx = $i + count($other->limbs);
            while ($carry > 0) {
                $cur = $out[$idx] + $carry;
                $out[$idx] = $cur % self::BASE;
                $carry = intdiv($cur, self::BASE);
                $idx++;
            }
        }
        return new self($this->sign * $other->sign, $out);
    }

    public function mulSmall(int $factor): self
    {
        if ($factor === 0 || $this->isZero()) {
            return self::zero();
        }
        if (abs($factor) >= self::BASE) {
            return $this->mul(self::fromInt($factor));
        }
        $sign = $factor < 0 ? -$this->sign : $this->sign;
        $factor = abs($factor);
        $out = [];
        $carry = 0;
        foreach ($this->limbs as $limb) {
            $cur = $limb * $factor + $carry;
            $out[] = $cur % self::BASE;
            $carry = intdiv($cur, self::BASE);
        }
        while ($carry > 0) {
            $out[] = $carry % self::BASE;
            $carry = intdiv($carry, self::BASE);
        }
        return new self($sign, $out);
    }

    public function square(): self
    {
        return $this->mul($this);
    }

    /** @return array{0:self,1:self} */
    public function divMod(self $divisor): array
    {
        if ($divisor->isZero()) {
            throw new InvalidArgumentException('DIVISION_BY_ZERO');
        }

        $d = $divisor->abs();
        $a = $this->abs();
        [$qAbs, $rAbs] = self::divModPositive($a, $d);

        if ($this->sign >= 0) {
            $q = $divisor->sign > 0 ? $qAbs : $qAbs->neg();
            return [$q, $rAbs];
        }

        if ($rAbs->isZero()) {
            $q = $divisor->sign > 0 ? $qAbs->neg() : $qAbs;
            return [$q, self::zero()];
        }

        $qFloorAbs = $qAbs->add(self::one());
        $q = $divisor->sign > 0 ? $qFloorAbs->neg() : $qFloorAbs;
        $r = $d->sub($rAbs);
        return [$q, $r];
    }

    public function floorDiv(self $divisor): self
    {
        return $this->divMod($divisor)[0];
    }

    public function mod(self $divisor): self
    {
        $r = $this->divMod($divisor)[1];
        if ($divisor->sign < 0 && !$r->isZero()) {
            return $r->neg();
        }
        return $r;
    }

    public function modSmall(int $divisor): int
    {
        if ($divisor <= 0) {
            throw new InvalidArgumentException('MOD_SMALL_DIVISOR');
        }
        $r = 0;
        for ($i = count($this->limbs) - 1; $i >= 0; $i--) {
            $r = (int)(($r * self::BASE + $this->limbs[$i]) % $divisor);
        }
        if ($this->sign >= 0 || $r === 0) {
            return $r;
        }
        return $divisor - $r;
    }

    public function toIntExact(): int
    {
        $s = $this->toString();
        if (!preg_match('/^-?\d{1,19}$/', $s)) {
            throw new RuntimeException('BIGINT_NOT_NATIVE_INT');
        }
        $v = filter_var($s, FILTER_VALIDATE_INT);
        if ($v === false) {
            throw new RuntimeException('BIGINT_NOT_NATIVE_INT');
        }
        return $v;
    }

    public function toString(): string
    {
        if ($this->sign === 0) {
            return '0';
        }
        $parts = array_reverse($this->limbs);
        $s = (string)array_shift($parts);
        foreach ($parts as $part) {
            $s .= str_pad((string)$part, self::BASE_DIGITS, '0', STR_PAD_LEFT);
        }
        return $this->sign < 0 ? '-' . $s : $s;
    }

    public function __toString(): string
    {
        return $this->toString();
    }

    /** @return array{0:self,1:self} */
    private static function divModPositive(self $a, self $d): array
    {
        if ($a->compare($d) < 0) {
            return [self::zero(), $a];
        }
        if (count($d->limbs) === 1) {
            $divisor = $d->limbs[0];
            $out = array_fill(0, count($a->limbs), 0);
            $rem = 0;
            for ($i = count($a->limbs) - 1; $i >= 0; $i--) {
                $cur = $rem * self::BASE + $a->limbs[$i];
                $out[$i] = intdiv($cur, $divisor);
                $rem = $cur % $divisor;
            }
            return [new self(1, $out), self::fromInt($rem)];
        }

        $qBigEndian = [];
        $remainder = self::zero();
        for ($i = count($a->limbs) - 1; $i >= 0; $i--) {
            $remainder = $remainder->shiftBaseAndAdd($a->limbs[$i]);
            $low = 0;
            $high = self::BASE - 1;
            $best = 0;
            while ($low <= $high) {
                $mid = intdiv($low + $high, 2);
                $probe = $d->mulSmall($mid);
                $cmp = $probe->compare($remainder);
                if ($cmp <= 0) {
                    $best = $mid;
                    $low = $mid + 1;
                } else {
                    $high = $mid - 1;
                }
            }
            $qBigEndian[] = $best;
            if ($best !== 0) {
                $remainder = $remainder->sub($d->mulSmall($best));
            }
        }

        $first = 0;
        while ($first < count($qBigEndian) - 1 && $qBigEndian[$first] === 0) {
            $first++;
        }
        $qBigEndian = array_slice($qBigEndian, $first);
        return [new self(1, array_reverse($qBigEndian)), $remainder];
    }

    private function shiftBaseAndAdd(int $limb): self
    {
        if ($this->isZero()) {
            return new self($limb === 0 ? 0 : 1, [$limb]);
        }
        $limbs = $this->limbs;
        array_unshift($limbs, $limb);
        return new self(1, $limbs);
    }

    /** @param list<int> $a @param list<int> $b */
    private static function compareAbsLimbs(array $a, array $b): int
    {
        if (count($a) !== count($b)) {
            return count($a) <=> count($b);
        }
        for ($i = count($a) - 1; $i >= 0; $i--) {
            if ($a[$i] !== $b[$i]) {
                return $a[$i] <=> $b[$i];
            }
        }
        return 0;
    }

    /** @param list<int> $a @param list<int> $b @return list<int> */
    private static function addAbsLimbs(array $a, array $b): array
    {
        $n = max(count($a), count($b));
        $out = [];
        $carry = 0;
        for ($i = 0; $i < $n; $i++) {
            $cur = ($a[$i] ?? 0) + ($b[$i] ?? 0) + $carry;
            if ($cur >= self::BASE) {
                $cur -= self::BASE;
                $carry = 1;
            } else {
                $carry = 0;
            }
            $out[] = $cur;
        }
        if ($carry !== 0) {
            $out[] = $carry;
        }
        return $out;
    }

    /** @param list<int> $a @param list<int> $b @return list<int> */
    private static function subAbsLimbs(array $a, array $b): array
    {
        $out = [];
        $borrow = 0;
        $n = count($a);
        for ($i = 0; $i < $n; $i++) {
            $cur = $a[$i] - ($b[$i] ?? 0) - $borrow;
            if ($cur < 0) {
                $cur += self::BASE;
                $borrow = 1;
            } else {
                $borrow = 0;
            }
            $out[] = $cur;
        }
        return $out;
    }
}

File: src/bigint.lua
local BigInt = {}
BigInt.__index = BigInt

local BASE = 10000000
local BASE_DIGITS = 7

local function is_big(x)
    return type(x) == "table" and getmetatable(x) == BigInt
end

local function normalize(x)
    local d = x.d
    while #d > 0 and d[#d] == 0 do
        d[#d] = nil
    end
    if #d == 0 then
        x.s = 0
    end
    return x
end

local function new(sign, digits)
    return normalize(setmetatable({s = sign, d = digits}, BigInt))
end

function BigInt.from(x)
    if is_big(x) then
        local d = {}
        for i = 1, #x.d do d[i] = x.d[i] end
        return new(x.s, d)
    end
    if type(x) == "number" then
        assert(math.type(x) == "integer", "BigInt mande yon nonb antye")
        if x == 0 then return new(0, {}) end
        local sign = x < 0 and -1 or 1
        if x < 0 then x = -x end
        local d = {}
        while x > 0 do
            d[#d + 1] = x % BASE
            x = x // BASE
        end
        return new(sign, d)
    end
    if type(x) == "string" then
        local sign = 1
        if x:sub(1,1) == "-" then
            sign = -1
            x = x:sub(2)
        elseif x:sub(1,1) == "+" then
            x = x:sub(2)
        end
        assert(x:match("^%d+$"), "BigInt mande yon chèn desimal")
        x = x:gsub("^0+", "")
        if x == "" then return new(0, {}) end
        local d = {}
        local i = #x
        while i > 0 do
            local j = math.max(1, i - BASE_DIGITS + 1)
            d[#d + 1] = tonumber(x:sub(j, i))
            i = j - 1
        end
        return new(sign, d)
    end
    error("Kalite sa a pa ka vin BigInt")
end

function BigInt:clone()
    return BigInt.from(self)
end

function BigInt:isZero()
    return self.s == 0
end

local function cmp_abs(a, b)
    if #a.d ~= #b.d then
        return #a.d < #b.d and -1 or 1
    end
    for i = #a.d, 1, -1 do
        if a.d[i] ~= b.d[i] then
            return a.d[i] < b.d[i] and -1 or 1
        end
    end
    return 0
end

function BigInt.compare(a, b)
    a, b = BigInt.from(a), BigInt.from(b)
    if a.s ~= b.s then return a.s < b.s and -1 or 1 end
    if a.s == 0 then return 0 end
    local c = cmp_abs(a, b)
    return a.s > 0 and c or -c
end

local function add_abs(a, b)
    local d, carry = {}, 0
    local n = math.max(#a.d, #b.d)
    for i = 1, n do
        local v = (a.d[i] or 0) + (b.d[i] or 0) + carry
        if v >= BASE then v, carry = v - BASE, 1 else carry = 0 end
        d[i] = v
    end
    if carry ~= 0 then d[n + 1] = carry end
    return d
end

local function sub_abs(a, b)
    local d, borrow = {}, 0
    for i = 1, #a.d do
        local v = a.d[i] - (b.d[i] or 0) - borrow
        if v < 0 then v, borrow = v + BASE, 1 else borrow = 0 end
        d[i] = v
    end
    return d
end

function BigInt.add(a, b)
    a, b = BigInt.from(a), BigInt.from(b)
    if a.s == 0 then return b end
    if b.s == 0 then return a end
    if a.s == b.s then return new(a.s, add_abs(a, b)) end
    local c = cmp_abs(a, b)
    if c == 0 then return new(0, {}) end
    if c > 0 then return new(a.s, sub_abs(a, b)) end
    return new(b.s, sub_abs(b, a))
end

function BigInt.neg(a)
    a = BigInt.from(a)
    if a.s == 0 then return a end
    a.s = -a.s
    return a
end

function BigInt.sub(a, b)
    return BigInt.add(a, BigInt.neg(b))
end

function BigInt.mulSmall(a, n)
    a = BigInt.from(a)
    assert(type(n) == "number" and math.type(n) == "integer", "Miltiplikatè a dwe antye")
    if a.s == 0 or n == 0 then return new(0, {}) end
    local sign = a.s
    if n < 0 then sign, n = -sign, -n end
    local d, carry = {}, 0
    for i = 1, #a.d do
        local v = a.d[i] * n + carry
        d[i] = v % BASE
        carry = v // BASE
    end
    while carry > 0 do
        d[#d + 1] = carry % BASE
        carry = carry // BASE
    end
    return new(sign, d)
end

function BigInt.mul(a, b)
    a, b = BigInt.from(a), BigInt.from(b)
    if a.s == 0 or b.s == 0 then return new(0, {}) end
    local d = {}
    for i = 1, #a.d + #b.d do d[i] = 0 end
    for i = 1, #a.d do
        local carry = 0
        for j = 1, #b.d do
            local k = i + j - 1
            local v = d[k] + a.d[i] * b.d[j] + carry
            d[k] = v % BASE
            carry = v // BASE
        end
        local k = i + #b.d
        while carry > 0 do
            local v = (d[k] or 0) + carry
            d[k] = v % BASE
            carry = v // BASE
            k = k + 1
        end
    end
    return new(a.s * b.s, d)
end

local function abs_value(a)
    a = BigInt.from(a)
    if a.s < 0 then a.s = 1 end
    return a
end

local function shift_base_add(a, limb)
    if a.s == 0 and limb == 0 then return new(0, {}) end
    local d = {limb}
    for i = 1, #a.d do d[i + 1] = a.d[i] end
    return new(1, d)
end

local function divmod_abs(a, b)
    assert(b.s > 0, "Divizè a dwe pozitif")
    if a.s == 0 then return new(0, {}), new(0, {}) end
    if cmp_abs(a, b) < 0 then return new(0, {}), a end
    local q_be = {}
    local r = new(0, {})
    for i = #a.d, 1, -1 do
        r = shift_base_add(r, a.d[i])
        local lo, hi, best = 0, BASE - 1, 0
        while lo <= hi do
            local mid = (lo + hi) // 2
            local prod = BigInt.mulSmall(b, mid)
            local c = cmp_abs(prod, r)
            if c <= 0 then
                best = mid
                lo = mid + 1
            else
                hi = mid - 1
            end
        end
        q_be[#q_be + 1] = best
        if best ~= 0 then r = BigInt.sub(r, BigInt.mulSmall(b, best)) end
    end
    local qd = {}
    for i = #q_be, 1, -1 do qd[#qd + 1] = q_be[i] end
    return new(1, qd), normalize(r)
end

function BigInt.divmodFloor(a, b)
    a, b = BigInt.from(a), BigInt.from(b)
    assert(b.s > 0, "Divizè a dwe yon BigInt pozitif")
    if a.s >= 0 then
        return divmod_abs(a, b)
    end
    local qa, ra = divmod_abs(abs_value(a), b)
    if ra.s == 0 then
        return BigInt.neg(qa), ra
    end
    return BigInt.neg(BigInt.add(qa, 1)), BigInt.sub(b, ra)
end

function BigInt.floorDiv(a, b)
    local q = BigInt.divmodFloor(a, b)
    return q
end

function BigInt.mod(a, b)
    local _, r = BigInt.divmodFloor(a, b)
    return r
end

function BigInt.pow(a, e)
    a = BigInt.from(a)
    assert(type(e) == "number" and math.type(e) == "integer" and e >= 0, "Eksponan an dwe antye epi pa negatif")
    local result = BigInt.from(1)
    local base = a
    while e > 0 do
        if e % 2 == 1 then result = BigInt.mul(result, base) end
        e = e // 2
        if e > 0 then base = BigInt.mul(base, base) end
    end
    return result
end

function BigInt.abs(a)
    return abs_value(a)
end

function BigInt.toNumber(a)
    a = BigInt.from(a)
    local n = 0
    for i = #a.d, 1, -1 do
        if n > (math.maxinteger - a.d[i]) // BASE then
            return nil
        end
        n = n * BASE + a.d[i]
    end
    return a.s < 0 and -n or n
end

function BigInt.__tostring(a)
    if a.s == 0 then return "0" end
    local parts = {}
    parts[1] = tostring(a.d[#a.d])
    for i = #a.d - 1, 1, -1 do
        parts[#parts + 1] = string.format("%0" .. BASE_DIGITS .. "d", a.d[i])
    end
    local s = table.concat(parts)
    return a.s < 0 and ("-" .. s) or s
end

function BigInt.__eq(a, b) return BigInt.compare(a, b) == 0 end
function BigInt.__lt(a, b) return BigInt.compare(a, b) < 0 end
function BigInt.__le(a, b) return BigInt.compare(a, b) <= 0 end
function BigInt.__unm(a) return BigInt.neg(a) end
function BigInt.__add(a, b) return BigInt.add(a, b) end
function BigInt.__sub(a, b) return BigInt.sub(a, b) end
function BigInt.__mul(a, b) return BigInt.mul(a, b) end
function BigInt.__idiv(a, b) return BigInt.floorDiv(a, b) end
function BigInt.__mod(a, b) return BigInt.mod(a, b) end
function BigInt.__pow(a, b) return BigInt.pow(a, b) end

BigInt.BASE = BASE
BigInt.ZERO = BigInt.from(0)
BigInt.ONE = BigInt.from(1)

return BigInt
